File: src/api/report.ts
import { invoke, PageResult } from "@/api";

export interface Report {
  id: number;
  type: string;
  title: string;
  content: string;
  start_date: string;
  end_date: string;
  commit_summary: string;
  state: number;
  created_at: string;
  updated_at: string;
}

export interface ReportBrief {
  id: number;
  title: string;
  start_date: string;
  end_date: string;
  created_at: string;
}

export interface MonthSummary {
  daily_reports: ReportBrief[];
  weekly_reports: ReportBrief[];
}

function readString(obj: Record<string, unknown>, snake: string, camel?: string): string {
  const value = obj[snake] ?? (camel ? obj[camel] : undefined);
  return typeof value === "string" ? value : "";
}

function readNumber(obj: Record<string, unknown>, key: string): number {
  const value = obj[key];
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

function normalizeReport(raw: unknown): Report {
  const obj = (raw ?? {}) as Record<string, unknown>;
  return {
    id: readNumber(obj, "id"),
    type: readString(obj, "type"),
    title: readString(obj, "title"),
    content: readString(obj, "content"),
    start_date: readString(obj, "start_date", "startDate"),
    end_date: readString(obj, "end_date", "endDate"),
    commit_summary: readString(obj, "commit_summary", "commitSummary"),
    state: readNumber(obj, "state"),
    created_at: readString(obj, "created_at", "createdAt"),
    updated_at: readString(obj, "updated_at", "updatedAt"),
  };
}

function normalizeReportBrief(raw: unknown): ReportBrief {
  const obj = (raw ?? {}) as Record<string, unknown>;
  return {
    id: readNumber(obj, "id"),
    title: readString(obj, "title"),
    start_date: readString(obj, "start_date", "startDate"),
    end_date: readString(obj, "end_date", "endDate"),
    created_at: readString(obj, "created_at", "createdAt"),
  };
}

function normalizeMonthSummary(raw: unknown): MonthSummary {
  const obj = (raw ?? {}) as Record<string, unknown>;
  const dailyRaw = obj.daily_reports ?? obj.dailyReports;
  const weeklyRaw = obj.weekly_reports ?? obj.weeklyReports;
  return {
    daily_reports: Array.isArray(dailyRaw) ? dailyRaw.map((item) => normalizeReportBrief(item)) : [],
    weekly_reports: Array.isArray(weeklyRaw) ? weeklyRaw.map((item) => normalizeReportBrief(item)) : [],
  };
}

function normalizeReportPage(raw: unknown): PageResult<Report> {
  const obj = (raw ?? {}) as Record<string, unknown>;
  const recordsRaw = obj.records;
  return {
    records: Array.isArray(recordsRaw) ? recordsRaw.map((item) => normalizeReport(item)) : [],
    total: readNumber(obj, "total"),
    page: readNumber(obj, "page"),
    size: readNumber(obj, "size"),
  };
}

export const reportApi = {
  list: (params?: { page?: number; size?: number; type?: string; keyword?: string }) =>
    invoke<unknown>("list_reports", { req: params || {} }).then(normalizeReportPage),
  detail: (id: number) =>
    invoke<unknown>("get_report_detail", { id }).then(normalizeReport),
  generate: (data: { type: string; start_date: string; end_date: string; force?: boolean; append_existing?: boolean; author_email?: string; project_ids?: number[] }) =>
    invoke<unknown>("generate_report", { req: data }).then(normalizeReport),
  update: (data: { id: number; title?: string; content?: string }) =>
    invoke<void>("update_report", { req: data }),
  delete: (id: number) =>
    invoke<void>("delete_report", { id }),
  monthSummary: (year: number, month: number) =>
    invoke<unknown>("get_month_summary", { req: { year, month } }).then(normalizeMonthSummary),
};

File: src-tauri/src/models/report.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Report {
    pub id: i32,
    pub r#type: String,
    pub title: String,
    pub content: String,
    pub start_date: String,
    pub end_date: String,
    pub commit_summary: String,
    pub state: i32,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ReportTemplate {
    pub id: i32,
    pub r#type: String,
    pub name: String,
    pub content: String,
    pub is_default: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ReportListReq {
    #[serde(default)]
    pub page: Option<i64>,
    #[serde(default)]
    pub size: Option<i64>,
    #[serde(default)]
    pub r#type: Option<String>,
    #[serde(default)]
    pub keyword: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ReportGenerateReq {
    pub r#type: String,
    pub start_date: String,
    pub end_date: String,
    #[serde(default)]
    pub force: bool,
    #[serde(default)]
    pub append_existing: bool,
    #[serde(default)]
    pub author_email: Option<String>,
    #[serde(default)]
    pub project_ids: Option<Vec<i32>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ReportUpdateReq {
    pub id: i32,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub content: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ReportMonthSummaryReq {
    pub year: i32,
    pub month: i32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct MonthSummaryRsp {
    pub daily_reports: Vec<ReportBrief>,
    pub weekly_reports: Vec<ReportBrief>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ReportBrief {
    pub id: i32,
    pub title: String,
    pub start_date: String,
    pub end_date: String,
    pub created_at: String,
}
